source/_posts/2024-08-16-state-of-generics-and-collections.md

```diff
@@ -0,0 +1,424 @@
+- Generic type information is not enforced at runtime (so if you don't run a static analysis tool in advance, they accomplish nothing).
+
+## Erased Generic Type Declarations
+
+Seeing the difficulties with implementing reified generics in PHP core, it has been proposed to implement them only syntactically, leaving the task of type checking to static analysers.
+
+In this alternative, the PHP syntax would be changed so that type declarations, class declarations, and function declarations accept generic syntax, but the PHP engine would not check them.
+
+We can call these “Erased” type declarations because the engine will simply ignore them at runtime. This alternative could be implemented in various ways:
+
+- As part of php-src
+- In an extension
+- In an autoloader
+- etc
+
+Here is what the same Dict class as above would look like:
+
+	class Dict<Key,Value>
+	{
+		public function __construct(private array<Key,Value> $entries) {}
+
+		public function set(Key $key, Value $value): self
+		{
+			$this->entries[$key] = $value;
+			return $this;
+		}
+	}
+
+	function f(Dict<Key,Value> $dict) {}
+
+	$dict = new Dict([1 => 'foo']);
+	$dict->set("foo", "bar"); // Static analyser error
+	$dict->set(1, "bar");     // Ok
+	f($dict);                 // Static analyser error
+
+This addresses the docblock verbosity issue of the Static Analysis alternative, but this introduces an inconsistency: Type declarations cause coercion, but erased generic type declarations will not.
+
+Consider the following example:
+
+	class StringList
+	{
+		public function add(string $value)
+		{
+			$this->values[] = $value;
+		}
+	}
+
+	class List<T>
+	{
+		public function add(T $value)
+		{
+			$this->values[] = $value;
+		}
+	}
+
+	$list = new StringList();
+	$list->add(123); // coerced to string
+
+	$list = new List<string>();
+	$list->add(123); // NOT coerced to string
+
+In this scenario, the first call to add() will coerce the argument to string, but not the second one.
+
+In languages like Java, which has erased generics on top of a traditional type system, the compiler does type checking, so inconsistencies like the above do not exist. However, in PHP these are unavoidable.
+
+Another drawback of erased generics is they are not visible at runtime. This would prevent pattern matching from seeing generic type arguments, for example.
+
+## Fully Erased Type Declarations
+
+One way to address the inconsistency issue of Erased Generics is to change all type declarations to be erased. This could be opt-in with a `declare()` statement:
+
+	declare(types=erased);
+
+In this alternative, the engine would stop checking types at runtime. In the previous example, both calls to add() would have the same behaviour: the value is not coerced. It is then up to the user to check types with an analyser.
+
+This is not uncommon in mainstream interpreted languages, as all of Javascript (via TypeScript), Python, and Ruby have fully erased type declarations.
+
```

Comment:
@arnaud-lb, @crell, @derickr, what do you think about adding some comment about the potential/feasibility for this in PHP?

Reply:
I think the biggest argument is already made: It creates inconsistency.  The second biggest is that we already have erased types, via PHPStan/Psalm, and moving them into the language syntax doesn't really buy us anything.

Reply:
I see now. For me it's not clear from the current text that full erased inherit the same problems. It seemed like a `declare(types=erased);` solves the inconsistency, and then the text suddenly breaks.

Reply:
> what do you think about adding some comment about the potential/feasibility for this in PHP?

In terms of feasibility, I think it's trivial (as long as we don't maintain/provide an analyzer). I don't know how much the idea would be accepted though.

> I think the biggest argument is already made: It creates inconsistency.

This argument is brought in the "Erased Generic Type Declarations" section, and this section ("Fully Erased Type Declarations") introduces a possible solution to this with  `declare(types=erased)`. There is no inconsistency anymore, at the cost of erasing all types.

>  The second biggest is that we already have erased types, via PHPStan/Psalm, and moving them into the language syntax doesn't really buy us anything

That's true, although we could see this as better ergonomics or reduced verbosity. This can make the language more pleasant to work with.

I suspect this is the primary reason for which users want generics, because they already have them in practice.

We could add a comment like "Letting users opt-in for fully erased typing and generics in a per-file basis would be a trivial change in core, and would bring PHPStan/Psalm-like generics to PHP with a better syntax".

Reply:
erased types should have a slight perf advantage even for current code, maybe worth mentioning?
I like them personally :) One drawback is that the syntax might not be as extensible as userland-implemented syntax, but I guess we might now have enough background to cover the domain? I'm just wondering about type extensions like `non-empty-string`, etc. For another discussion I guess :)

Reply:
This totally opens the door to types like `non-empty-string`. Implementation would be trivial if we recognize a finite set of types, but probably more difficult if we want something more abstract like https://github.com/tc39/proposal-type-annotations.

We could add: 

```
Some additional benefits of this solutions are:
- The lack of type checking at runtime in opted-in code would result in a short term performance improvement
- It opens the door to extensions of the type system, with types like `non-empty-string`, `list`, `int<range>`, `class-string`, conditional types, etc.

Reply:
Would erased types cause issues for reflection?

Reply:
> Would erased types cause issues for reflection?

Probably not. I think we should treat these types as opaque strings and store them internally. Reflection can either expose them as strings via a separate method, or as a new ReflectionType subclass.

> I'm unclear how erased types makes other one-off pseudo-types easier, other than "well we just ignore them anyway do put whatever the F you want as a type."

Yes it only makes the syntax nicer. (Which is something users may care about.)

Reply:
I mean full type erasure.  I don't know what the ideal reflection/generics behavior is, but I accept that it's weird.  But full type erasure sounds like it would break reflection, especially for those of us that are doing fancy things with it. :smile:  (ORMs, serialization, etc.)